src/components/activities/RosterView.tsx

```diff
@@ -0,0 +1,200 @@
+              </TableCell>
+              <TableCell align="center" sx={headerCellStyle}>
+                Sign Out
+              </TableCell>
+            </TableRow>
+          </TableHead>
+          <TableBody>
+            {roster.entries.map((entry, i) => (
+              <RosterRow key={i} rosterEntry={entry} />
+            ))}
+          </TableBody>
+        </Table>
+      </Paper>
+    </ToolbarPage>
+  );
+}
+
+function RosterRow({ rosterEntry }: { rosterEntry: RosterEntry }) {
+  return (
+    <TableRow>
+      <TableCell size="small">{rosterEntry.participantName}</TableCell>
+      {Object.values(rosterEntry.timestamps).map((time, i) => (
+        <TableCell key={i} align="center" size="small">
+          <Stack>{time ? <RosterTimeValue time={time} /> : undefined}</Stack>
+        </TableCell>
+      ))}
+    </TableRow>
+  );
+}
+
+function RosterTimeValue({ time }: { time: number }) {
+  const dateString = new Date(time).toLocaleString('en-US', { month: '2-digit', day: '2-digit', year: '2-digit' });
+  const timeString = new Date(time).toLocaleString('en-US', { hour12: false, hour: '2-digit', minute: '2-digit' }).replace(':', '');
+  return (
+    <>
+      <Typography variant="h6">{timeString}</Typography>
+      <Typography variant="caption">{dateString}</Typography>
+    </>
+  );
+}
+
+enum RosterStage {
+  NA = 0,
+  SignIn = 1,
+  ArriveBase = 2,
+  DepartBase = 3,
+  SignOut = 4,
+}
+
+const rosterStages: Record<ParticipantStatus, RosterStage> = {
+  [ParticipantStatus.NotResponding]: RosterStage.NA,
+  [ParticipantStatus.Standby]: RosterStage.NA,
+  [ParticipantStatus.Assigned]: RosterStage.NA,
+  [ParticipantStatus.SignedIn]: RosterStage.SignIn,
+  [ParticipantStatus.Remote]: RosterStage.SignIn,
+  [ParticipantStatus.Available]: RosterStage.ArriveBase,
+  [ParticipantStatus.Demobilized]: RosterStage.DepartBase,
+  [ParticipantStatus.SignedOut]: RosterStage.SignOut,
+};
+
+interface IRosterEntry {
+  participantId: string;
+  participantName: string;
+  timestamps: {
+    [RosterStage.SignIn]: number;
+    [RosterStage.ArriveBase]: number;
+    [RosterStage.DepartBase]: number;
+    [RosterStage.SignOut]: number;
+  };
+  miles: number;
+}
+
+const scrubTimeline = (timeline: Array<ParticipantUpdate>): Array<ParticipantUpdate> => {
+  const newTimeline = [];
+  let priorStage = 0;
+  for (let i = timeline.length - 1; i >= 0; i--) {
+    const t = timeline[i];
+    const stage: RosterStage = rosterStages[t.status] ?? undefined;
+    if (stage === RosterStage.NA) {
+      if (i === 0) {
+        newTimeline.unshift(t); // Keep if latest status
+        priorStage = stage;
+      }
+      continue;
+    }
+    if (stage === RosterStage.SignOut) {
+      newTimeline.unshift(t);
+      priorStage = RosterStage.NA;
+      continue;
+    }
+    if (stage === priorStage + 1) {
+      newTimeline.unshift(t);
+      priorStage = stage;
+      continue;
+    }
+  }
+  return newTimeline;
+};
+
+class Roster {
```

Comment:
Coming from an OOP background I loathe the functional approach. Pure spagett. It takes forever to work out how things are happening because the code provides no context.

I'll align with the rest of the project, here. But, you will never convince me functional better.

Reply:
Not saying I disagree lol. It's an experience, that's for sure. But my primary style is to err on the side of consistency.